### ISSUE TITLE
企业微信审批回调监听消息，解析后审批人信息丢失

### ISSUE THREAD

**Issue by leone-007** [Wechat-Group/WxJava#1888]
### 简要描述
使用消息事件监听，监听审批事件变化，未正确解析到审批人信息

### 模块版本情况
* WxJava 模块名: weixin-java-cp
* WxJava 版本号:3.9.0

### 详细描述
使用消息事件监听，监听审批事件变化，使用   wxMessage.getAllFieldsMap()，从map信息里获取的消息allFieldsMap下SpRecord里Details下Approver未正确解析到数据。
"allFieldsMap" : {
    "ApprovalInfo" : {
      "SpStatus" : "1",
      "ApplyTime" : "1606220519",
      "StatuChangeEvent" : "2",
      "SpRecord" : {
        "SpStatus" : "1",
        "ApproverAttr" : "1",
        "Details" : {
          "SpStatus" : "1",
          "Speech" : "",
          "Approver" : "",
          "SpTime" : "0"
        }
      },

**Comment by binarywang** [comment]
原始报文里有吗？发一下对照下？

**Comment by leone-007** [comment]
接收微信请求：`[signature=[cde8cc67f19818024102e6ba3e6d7201bc0ac291], timestamp=[1606268627], nonce=[1606277571], requestBody=[
<xml><ToUserName><![CDATA[wwead3238409715ded]]></ToUserName><Encrypt><![CDATA[8i49ajaCceli50eNNazgHGGF4OrSabJy2xHypaMrTnsUnmSRpPxQ1Je19SopPjb/okAb1NvS2KRwbHvNOiP/PE21xCNkF5Zec/6viutjq1KlOTL/UcvpyOOOMKnzTzOw2eEhPx3axpRnWFbQkqbl9E4eUbvyA85B77B95G3BgdJ8vrGv5umkMo31XsdB+OZTnZLK8pG/iYbJz644SZKGSpx5yEztnyQvUSUhqOwVAqxFhdBb8IA5iw/BpnJCkcYl/zY8w/y9gXK0aq1045eSGs8pu8EvErqJFPQ6tLHYRfMvGT7aCna98cpOl4pHs3WFC25NWoiDrj1jbqAM883J4YkbkYTe+vpuVsumjZ9uBbQQRMyAcNTrUZ/5b9U3DyvN2QGEDP/xJF+rJ8uX+mc+7oj72Y4q3SbuQw1bVaxi8EHJkLppZ26RKkx3THT2UXahTDNm9sjZzaOZuvyPsuvb8ggmWKlw3lFS6gqQYHueM7urnDlyf9xnLZhdwy0x4vdFDxpZFVEFFJMupQEEgZXlDqc9HCGwU39K1ImvKJiMdOSZdJ91P2trFhukEa8UkNYLKzWFVDXLC8XNHEIcuf8XyNExPtcdwjNhDvSBXdUG/XV0eHruJPiSMwtYUcTX+9zkSHnAtFxHo2PbfV7fdsMOKT846crSwY30Pf62PyAg4a2oHqyiqI3OX5T0mYd0feAT9nne1Ds1ZftWeknmJAZySFQN1eBGOfPRw4ZKvQankjB22I80fmLkJeTD9idkT1LIIbpb0TACtQY4cK9K77tGyFBFy2yWpqwn21p7LR0e13e0gj9c8jvJtN3i4bzq7sqcZF0am3TcKm3RurIUxFjbPW7BOfAIvh/r4bftIKME3GEahwJsvuSX4JWrjHHiv2dkcKNW7FWpetHo3rXzSljq7GTJ3qb2sTtEBW6vN1X59R/l5wHVASPPSxcQfpK0ykH4lor6cBVJTFxe6d5a47dyeWfLxUevWJa4uo+3v5d+/MASkz4qPzMpHh0kw7+tsnIUdhZbDYyoP4llX0px1hCUk2+0/OQOjT1Hc8uCyUuiM9P2jPEo2SJ7qo7uVBTo7csvIPxbjJdOfz8mnmcddqhCDmQs7jRxiLiHfSTdeE6hUThXHFLuI/5ZUrZDguGs1ydeBbQ5rAyRiNbdeqLk9S9kJXFGwDDCppqpL+j2zDn9FTAzlWbBCfy0o3VNgCnUlZbWxZT6UPGQEYsF4nnkNkNYgzFCiMZ1+42as7bu1oiroLpgJCenZl4CyskvMM9Ef8sVTe6L4xdmsBbViDvvXA5axaG6VVh7yZXlOCImi4v3JJy3Lvqmp+WSe4cG6lX9xqNMWGfIud4svSgck6inNwWfSsE14fo6cauPHXBA8Bk6d3KekdclB1y/Ry0dvAC4DfMvgiWbRgf2QM9wn30DWnQr/DivygqJJ9RIV+IBxVfJiRA=]]></Encrypt><AgentID><![CDATA[3010040]]></AgentID></xml>
]`

**Comment by leone-007** [comment]
接收微信请求：[signature=[cde8cc67f19818024102e6ba3e6d7201bc0ac291], timestamp=[1606268627], nonce=[1606277571], requestBody=[
<xml><ToUserName><![CDATA[wwead3238409715ded]]></ToUserName><Encrypt><![CDATA[8i49ajaCceli50eNNazgHGGF4OrSabJy2xHypaMrTnsUnmSRpPxQ1Je19SopPjb/okAb1NvS2KRwbHvNOiP/PE21xCNkF5Zec/6viutjq1KlOTL/UcvpyOOOMKnzTzOw2eEhPx3axpRnWFbQkqbl9E4eUbvyA85B77B95G3BgdJ8vrGv5umkMo31XsdB+OZTnZLK8pG/iYbJz644SZKGSpx5yEztnyQvUSUhqOwVAqxFhdBb8IA5iw/BpnJCkcYl/zY8w/y9gXK0aq1045eSGs8pu8EvErqJFPQ6tLHYRfMvGT7aCna98cpOl4pHs3WFC25NWoiDrj1jbqAM883J4YkbkYTe+vpuVsumjZ9uBbQQRMyAcNTrUZ/5b9U3DyvN2QGEDP/xJF+rJ8uX+mc+7oj72Y4q3SbuQw1bVaxi8EHJkLppZ26RKkx3THT2UXahTDNm9sjZzaOZuvyPsuvb8ggmWKlw3lFS6gqQYHueM7urnDlyf9xnLZhdwy0x4vdFDxpZFVEFFJMupQEEgZXlDqc9HCGwU39K1ImvKJiMdOSZdJ91P2trFhukEa8UkNYLKzWFVDXLC8XNHEIcuf8XyNExPtcdwjNhDvSBXdUG/XV0eHruJPiSMwtYUcTX+9zkSHnAtFxHo2PbfV7fdsMOKT846crSwY30Pf62PyAg4a2oHqyiqI3OX5T0mYd0feAT9nne1Ds1ZftWeknmJAZySFQN1eBGOfPRw4ZKvQankjB22I80fmLkJeTD9idkT1LIIbpb0TACtQY4cK9K77tGyFBFy2yWpqwn21p7LR0e13e0gj9c8jvJtN3i4bzq7sqcZF0am3TcKm3RurIUxFjbPW7BOfAIvh/r4bftIKME3GEahwJsvuSX4JWrjHHiv2dkcKNW7FWpetHo3rXzSljq7GTJ3qb2sTtEBW6vN1X59R/l5wHVASPPSxcQfpK0ykH4lor6cBVJTFxe6d5a47dyeWfLxUevWJa4uo+3v5d+/MASkz4qPzMpHh0kw7+tsnIUdhZbDYyoP4llX0px1hCUk2+0/OQOjT1Hc8uCyUuiM9P2jPEo2SJ7qo7uVBTo7csvIPxbjJdOfz8mnmcddqhCDmQs7jRxiLiHfSTdeE6hUThXHFLuI/5ZUrZDguGs1ydeBbQ5rAyRiNbdeqLk9S9kJXFGwDDCppqpL+j2zDn9FTAzlWbBCfy0o3VNgCnUlZbWxZT6UPGQEYsF4nnkNkNYgzFCiMZ1+42as7bu1oiroLpgJCenZl4CyskvMM9Ef8sVTe6L4xdmsBbViDvvXA5axaG6VVh7yZXlOCImi4v3JJy3Lvqmp+WSe4cG6lX9xqNMWGfIud4svSgck6inNwWfSsE14fo6cauPHXBA8Bk6d3KekdclB1y/Ry0dvAC4DfMvgiWbRgf2QM9wn30DWnQr/DivygqJJ9RIV+IBxVfJiRA=]]></Encrypt><AgentID><![CDATA[3010040]]></AgentID></xml>
]

**Comment by binarywang** [comment]
你要把解密后的报文拿出来进行对比，自己先对比下看看？

**Comment by leone-007** [comment]
原始文件解密后确实有审批人信息，`<?xml version="1.0" encoding="utf-8"?>

<xml>
  <ToUserName><![CDATA[wwead3238409715ded]]></ToUserName>
  <FromUserName><![CDATA[sys]]></FromUserName>
  <CreateTime>1606268627</CreateTime>
  <MsgType><![CDATA[event]]></MsgType>
  <Event><![CDATA[sys_approval_change]]></Event>
  <AgentID>3010040</AgentID>
  <ApprovalInfo>
    <SpNo>202011240045</SpNo>
    <SpName><![CDATA[费用报销]]></SpName>
    <SpStatus>1</SpStatus>
    <TemplateId><![CDATA[ZLyca9D2cvnAB14xkYEtyE11hcGJidiFThJi5d]]></TemplateId>
    <ApplyTime>1606220519</ApplyTime>
    <Applyer>
      <UserId><![CDATA[ZhuLinYang]]></UserId>
      <Party><![CDATA[2]]></Party>
    </Applyer>
    <SpRecord>
      <SpStatus>2</SpStatus>
      <ApproverAttr>1</ApproverAttr>
      <Details>
        <Approver>
          <UserId><![CDATA[GaoYu]]></UserId>
        </Approver>
        <Speech><![CDATA[]]></Speech>
        <SpStatus>2</SpStatus>
        <SpTime>1606268626</SpTime>
      </Details>
    </SpRecord>
    <SpRecord>
      <SpStatus>1</SpStatus>
      <ApproverAttr>1</ApproverAttr>
      <Details>
        <Approver>
          <UserId><![CDATA[ZhuLinYang]]></UserId>
        </Approver>
        <Speech><![CDATA[]]></Speech>
        <SpStatus>1</SpStatus>
        <SpTime>0</SpTime>
      </Details>
    </SpRecord>
    <StatuChangeEvent>2</StatuChangeEvent>
  </ApprovalInfo>
</xml>
`

**Comment by leone-007** [comment]
使用企业微信组件解析出来的结果:`{
  "allFieldsMap" : {
    "ApprovalInfo" : {
      "SpStatus" : "1",
      "ApplyTime" : "1606220519",
      "StatuChangeEvent" : "2",
      "SpRecord" : {
        "SpStatus" : "1",
        "ApproverAttr" : "1",
        "Details" : {
          "SpStatus" : "1",
          "Speech" : "",
          "Approver" : "",
          "SpTime" : "0"
        }
      },
      "SpNo" : "202011240045",
      "SpName" : "费用报销",
      "Applyer" : {
        "Party" : "2",
        "UserId" : "ZhuLinYang"
      },
      "TemplateId" : "ZLyca9D2cvnAB14xkYEtyE11hcGJidiFThJi5d"
    },
    "CreateTime" : "1606268627",
    "Event" : "sys_approval_change",
    "ToUserName" : "wwead3238409715ded",
    "FromUserName" : "sys",
    "MsgType" : "event",
    "AgentID" : "1000012"
  },
  "agentId" : 1000012,
  "toUserName" : "wwead3238409715ded",
  "fromUserName" : "sys",
  "createTime" : 1606268627,
  "msgType" : "event",
  "event" : "sys_approval_change",
  "extAttrs" : {
    "items" : [ ]
  },
  "scanCodeInfo" : { },
  "sendPicsInfo" : {
    "picList" : [ ]
  },
  "sendLocationInfo" : { },
  "approvalInfo" : {
    "spNo" : "202011240045",
    "spName" : "费用报销",
    "spStatus" : 1,
    "templateId" : "ZLyca9D2cvnAB14xkYEtyE11hcGJidiFThJi5d",
    "applyTime" : 1606220519,
    "applier" : {
      "party" : "2"
    },
    "statusChangeEvent" : 2
  }
}`

**Comment by binarywang** [comment]
你要使用三个` 引用报文才容易看

**Comment by leone-007** [comment]
原始报文解析结果：
```{
  "xml": {
    "ToUserName": "wwead3238409715ded",
    "FromUserName": "sys",
    "CreateTime": "1606268627",
    "MsgType": "event",
    "Event": "sys_approval_change",
    "AgentID": "3010040",
    "ApprovalInfo": {
      "SpNo": "202011240045",
      "SpName": "费用报销",
      "SpStatus": "1",
      "TemplateId": "ZLyca9D2cvnAB14xkYEtyE11hcGJidiFThJi5d",
      "ApplyTime": "1606220519",
      "Applyer": {
        "UserId": "ZhuLinYang",
        "Party": "2"
      },
      "SpRecord": [
        {
          "SpStatus": "2",
          "ApproverAttr": "1",
          "Details": {
            "Approver": { "UserId": "GaoYu" },
            "SpStatus": "2",
            "SpTime": "1606268626"
          }
        },
        {
          "SpStatus": "1",
          "ApproverAttr": "1",
          "Details": {
            "Approver": { "UserId": "ZhuLinYang" },
            "SpStatus": "1",
            "SpTime": "0"
          }
        }
      ],
      "StatuChangeEvent": "2"
    }
  }
}```

**Comment by leone-007** [comment]
企业微信组件解析报文结果：

```{
  "allFieldsMap" : {
    "ApprovalInfo" : {
      "SpStatus" : "1",
      "ApplyTime" : "1606220519",
      "StatuChangeEvent" : "2",
      "SpRecord" : {
        "SpStatus" : "1",
        "ApproverAttr" : "1",
        "Details" : {
          "SpStatus" : "1",
          "Speech" : "",
          "Approver" : "",
          "SpTime" : "0"
        }
      },
      "SpNo" : "202011240045",
      "SpName" : "费用报销",
      "Applyer" : {
        "Party" : "2",
        "UserId" : "ZhuLinYang"
      },
      "TemplateId" : "ZLyca9D2cvnAB14xkYEtyE11hcGJidiFThJi5d"
    },
    "CreateTime" : "1606268627",
    "Event" : "sys_approval_change",
    "ToUserName" : "wwead3238409715ded",
    "FromUserName" : "sys",
    "MsgType" : "event",
    "AgentID" : "1000012"
  },
  "agentId" : 1000012,
  "toUserName" : "wwead3238409715ded",
  "fromUserName" : "sys",
  "createTime" : 1606268627,
  "msgType" : "event",
  "event" : "sys_approval_change",
  "extAttrs" : {
    "items" : [ ]
  },
  "scanCodeInfo" : { },
  "sendPicsInfo" : {
    "picList" : [ ]
  },
  "sendLocationInfo" : { },
  "approvalInfo" : {
    "spNo" : "202011240045",
    "spName" : "费用报销",
    "spStatus" : 1,
    "templateId" : "ZLyca9D2cvnAB14xkYEtyE11hcGJidiFThJi5d",
    "applyTime" : 1606220519,
    "applier" : {
      "party" : "2"
    },
    "statusChangeEvent" : 2
  }
}```

**Comment by binarywang** [comment]
醉了，在原来的那个上面改不就行了吗？发了这么多

**Comment by leone-007** [comment]
其中Details下信息未正确解析

**Comment by leone-007** [comment]
> 醉了，在原来的那个上面改不就行了吗？发了这么多

￣□￣｜｜

**Comment by Gyv12345** [comment]
这是当时学生写的，我给挪过来了，居然少了

**Comment by binarywang** [comment]
> 原始报文解析结果：
> 
> ```
>   "xml": {
>     "ToUserName": "wwead3238409715ded",
>     "FromUserName": "sys",
>     "CreateTime": "1606268627",
>     "MsgType": "event",
>     "Event": "sys_approval_change",
>     "AgentID": "3010040",
>     "ApprovalInfo": {
>       "SpNo": "202011240045",
>       "SpName": "费用报销",
>       "SpStatus": "1",
>       "TemplateId": "ZLyca9D2cvnAB14xkYEtyE11hcGJidiFThJi5d",
>       "ApplyTime": "1606220519",
>       "Applyer": {
>         "UserId": "ZhuLinYang",
>         "Party": "2"
>       },
>       "SpRecord": [
>         {
>           "SpStatus": "2",
>           "ApproverAttr": "1",
>           "Details": {
>             "Approver": { "UserId": "GaoYu" },
>             "SpStatus": "2",
>             "SpTime": "1606268626"
>           }
>         },
>         {
>           "SpStatus": "1",
>           "ApproverAttr": "1",
>           "Details": {
>             "Approver": { "UserId": "ZhuLinYang" },
>             "SpStatus": "1",
>             "SpTime": "0"
>           }
>         }
>       ],
>       "StatuChangeEvent": "2"
>     }
>   }
> }```
> ```

才发现个问题，为啥你这个不是xml呢？并不是微信的原始报文吧？

**Comment by leone-007** [comment]
这个我手动处理转换成json了，内容都是完整的

**Comment by leone-007** [comment]
> > 原始报文解析结果：
> > ```
> >   "xml": {
> >     "ToUserName": "wwead3238409715ded",
> >     "FromUserName": "sys",
> >     "CreateTime": "1606268627",
> >     "MsgType": "event",
> >     "Event": "sys_approval_change",
> >     "AgentID": "3010040",
> >     "ApprovalInfo": {
> >       "SpNo": "202011240045",
> >       "SpName": "费用报销",
> >       "SpStatus": "1",
> >       "TemplateId": "ZLyca9D2cvnAB14xkYEtyE11hcGJidiFThJi5d",
> >       "ApplyTime": "1606220519",
> >       "Applyer": {
> >         "UserId": "ZhuLinYang",
> >         "Party": "2"
> >       },
> >       "SpRecord": [
> >         {
> >           "SpStatus": "2",
> >           "ApproverAttr": "1",
> >           "Details": {
> >             "Approver": { "UserId": "GaoYu" },
> >             "SpStatus": "2",
> >             "SpTime": "1606268626"
> >           }
> >         },
> >         {
> >           "SpStatus": "1",
> >           "ApproverAttr": "1",
> >           "Details": {
> >             "Approver": { "UserId": "ZhuLinYang" },
> >             "SpStatus": "1",
> >             "SpTime": "0"
> >           }
> >         }
> >       ],
> >       "StatuChangeEvent": "2"
> >     }
> >   }
> > }```
> > ```
> 
> 才发现个问题，为啥你这个不是xml呢？不是微信的原始报文吧？

这个我手动处理转换成json了，内容都是完整的

**Comment by binarywang** [comment]
如果可以的话，你把原始xml发一下吧，方便写单元测试代码进行调试

**Comment by binarywang** [comment]
楼主呢？如果不方便就算了，暂时关闭
